### PR TITLE
svg_loader: <image> tag introduced

### PR DESCRIPTION
**Pull request by mmaciola**
This patch introduces embeded `<image>` tag for svg files.
Images can be now loaded from local path or using data URI Scheme (RFC2397).
Base64 and UTF8/URI encoded data are welcome.

example svg file:
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 200 200">
    <defs>
        <clipPath id="circleView">
            <circle cx="100" cy="100" r="40" fill="#FFFFFF" />
        </clipPath>
    </defs>
    <image href="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100'%3E%3Ccircle cx='50' cy='50' r='48' fill='none' stroke='%23000'/%3E%3Cpath d='M50,2a48,48 0 1 1 0,96a24 24 0 1 1 0-48a24 24 0 1 0 0-48'/%3E%3Ccircle cx='50' cy='26' r='6'/%3E%3Ccircle cx='50' cy='74' r='6' fill='%23FFF'/%3E%3C/svg%3E%0A" height="100" width="100" x="0" y="0" />
    <image href="data:image/svg+xml;utf8,%3Csvg+xmlns%3D%27http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%27+viewBox%3D%270+0+100+100%27%3E%3Ccircle+cx%3D%2750%27+cy%3D%2750%27+r%3D%2748%27+fill%3D%27none%27+stroke%3D%27%23000%27%2F%3E%3Cpath+d%3D%27M50%2C2a48%2C48+0+1+1+0%2C96a24+24+0+1+1+0-48a24+24+0+1+0+0-48%27%2F%3E%3Ccircle+cx%3D%2750%27+cy%3D%2726%27+r%3D%276%27%2F%3E%3Ccircle+cx%3D%2750%27+cy%3D%2774%27+r%3D%276%27+fill%3D%27%23FFF%27%2F%3E%3C%2Fsvg%3E%0D%0A" height="100" width="100" x="100" y="0" />
    <image href="data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHZpZXdCb3g9JzAgMCAxMDAgMTAwJz48Y2lyY2xlIGN4PSc1MCcgY3k9JzUwJyByPSc0OCcgZmlsbD0nbm9uZScgc3Ryb2tlPScjMDAwJy8+PHBhdGggZD0nTTUwLDJhNDgsNDggMCAxIDEgMCw5NmEyNCAyNCAwIDEgMSAwLTQ4YTI0IDI0IDAgMSAwIDAtNDgnLz48Y2lyY2xlIGN4PSc1MCcgY3k9JzI2JyByPSc2Jy8+PGNpcmNsZSBjeD0nNTAnIGN5PSc3NCcgcj0nNicgZmlsbD0nI0ZGRicvPjwvc3ZnPg0K" height="100" width="100" x="0" y="100" />
    <image href="src/examples/images/yinyang.svg" height="100" width="100" x="100" y="100" />
    <image href="src/examples/images/logo.png" height="80" width="80" x="60" y="60" clip-path="url(#circleView)" />
</svg>
```
Screenshot from 2021-07-01 10-48-33

@issue: #518

### CODE REVIEW

File: src/loaders/svg/tvgSvgSceneBuilder.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:368]: please do not include unwilling feature code. A patch must include only necessaries.

File: src/loaders/svg/tvgSvgSceneBuilder.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:371]: I think it's bad that SVG has PNG loader dependency. Thorvg provides Picture which abstracts them. So just use picture  is enough to go.

**Review by hermet — CHANGES_REQUESTED** [review]
Please check comments.

File: src/loaders/svg/tvgSvgSceneBuilder.cpp
Comment (mmaciola) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp]: removed `#ifdef`

File: src/loaders/svg/tvgSvgSceneBuilder.cpp
Comment (mmaciola) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp]: removed `#ifdef`

File: src/loaders/svg/tvgSvgLoaderCommon.h
Comment (hermet) [inline src/loaders/svg/tvgSvgLoaderCommon.h:162]: One line please.
float x, y, w, h;

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:27]: jfyi, when you introduce a new header, please compare the symbol size (before/after). If it does, please consider not to use if its possible. Many headers generate codes even just including.

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:262]: recommend to use "auto"

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:263]: length == 0

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:286]: unsigned int => auto

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:45]: func()
{
}

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:39]: func()
{
}

File: src/loaders/svg/tvgSvgUtil.cpp
Comment (hermet) [inline src/loaders/svg/tvgSvgUtil.cpp:293]: uint8_t => auto

**Review by hermet — CHANGES_REQUESTED** [review]
Please check comments.

File: src/loaders/svg/tvgSvgLoaderCommon.h
Comment (mmaciola) [reply src/loaders/svg/tvgSvgLoaderCommon.h:162]: one line there would be odd. File contain multiple structs definitions in the same manner.

File: src/loaders/svg/tvgSvgSceneBuilder.cpp
Comment (JSUYA) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:436]: The SVG file for verification of issue https://github.com/Samsung/thorvg/issues/457 (SVG.zip/e-image-033.svg) is testing recursive svg image loading.
A crash may occur in this case.

until this issue is resolved... it might be better to decide that svg image loading is not supported.

Comment (mmaciola) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp:440]: Please find the change. I added the code that checks the extension and disables the embedded svg loading.

File: src/loaders/svg/tvgSvgLoaderCommon.h
Comment (hermet) [reply src/loaders/svg/tvgSvgLoaderCommon.h:162]: @mmaciola I'd like to guide you a convention. See  other files. ex) tvgSwCommon.h

**Review by hermet — CHANGES_REQUESTED** [review]

**Review by JSUYA — COMMENTED** [review]
Could you check this svg case?
I'm not sure if the encoding is wrong or the data is wrong.
However, according to this PR, it seems that hrefs starting with `data:image/png;base64` should be displayed normally.

SVG/e-image-014.svg 
```html
<svg id="svg1" viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg"                        
     xmlns:xlink="http://www.w3.org/1999/xlink">
    <title>preserveAspectRatio='xMinYMin slice'</title>

    <image id="image1" x="36" y="3" width="128" height="64" xlink:href="data:image/png;base64,
iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAB3RJTUUH4gMLDwAjrsLbtwAAAAlw
SFlzAAAuIwAALiMBeKU/dgAAABl0RVh0Q29tbWVudABDcmVhdGVkIHdpdGggR0lNUFeBDhcAAAAG
UExURQAA/xjQP14JpdQAAAABYktHRACIBR1IAAAAFklEQVR42mMAgvp/IJTAhgdB1ADVAgDvdAnx
N1Ib1gAAAABJRU5ErkJggg=="
           preserveAspectRatio="xMinYMin slice"/>
    <rect id="rect1" x="36" y="3" width="128" height="64" fill="none" stroke="black"/>

    <image id="image2" x="68" y="70" width="64" height="128" xlink:href="data:image/png;base64,
iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAB3RJTUUH4gMLDwAjrsLbtwAAAAlw
SFlzAAAuIwAALiMBeKU/dgAAABl0RVh0Q29tbWVudABDcmVhdGVkIHdpdGggR0lNUFeBDhcAAAAG
UExURQAA/xjQP14JpdQAAAABYktHRACIBR1IAAAAFklEQVR42mMAgvp/IJTAhgdB1ADVAgDvdAnx
N1Ib1gAAAABJRU5ErkJggg=="
           preserveAspectRatio="xMinYMin slice"/>
    <rect id="rect2" x="68" y="70" width="64" height="128" fill="none" stroke="black"/>

    <!-- image frame -->
    <rect id="frame" x="1" y="1" width="198" height="198" fill="none" stroke="black"/>
</svg>
                                                                                               

```

**Comment by mmaciola** [comment]
@JSUYA Thanks for pointing. There was a problem with new line breaks. Now fixed. (note PR #570 is needed for png)
There is still `preserveAspectRatio` attribute not supported. Will be added in another pr in future.

**Review by JSUYA — APPROVED** [review]
LGTM
I don't think there are any issues with this patch. I wait for hermet's approval.

**Pull request merged**
